20250117-actors-pubsub.md

```diff
@@ -0,0 +1,140 @@
+# Actor PubSub Implementation Proposal
+
+- Author(s): @joshvanl
+
+## Overview
+
+This proposal introduces PubSub capabilities for Dapr Actors, enabling actors to subscribe to messages within their type for a particular topic.
+Additionally, actors of a specific type will have the ability to publish messages to a topic.
+A single designated PubSub component will be marked as the actor PubSub, similar to how an actor state store is defined.
+There can be only one actor PubSub state store per Dapr deployment.
+
+## Background
+
+Currently, Dapr provides PubSub functionality for applications but lacks native PubSub integration for actors.
+This limitation means that actors cannot communicate efficiently via publish-subscribe messaging patterns within their type.
+By enabling Actor PubSub, actors can directly subscribe to and publish messages to specific topics, improving scalability and communication efficiency.
+
+## Implementation Details
+
+### Actor PubSub Registration
+
+A single PubSub component will be designated as the actor PubSub, similar to how an actor state store is configured.
+The metadata configuration will follow the existing pattern:
+
+```yaml
+metadata:
+  - name: actorpubsub
+    value: "true"
+```
+
+Only one PubSub component can be marked as the actor PubSub.
+
+### Actor Subscriptions
+
+Actors will be able to subscribe to messages within their type for a particular topic.
+Subscriptions will be dynamically declared, and messages for that topic will be routed through the established bi-directional gRPC stream.
+
+Upon deactivation, either during shutdown or actor rebalancing, the subscription to all topics by the actor will be removed.
```

Comment:
It is not clear how rebalancing will work with the subscription. If the subscription is per actor type, how do you handle redirecting the message to the right actor instance?

Also, how to handle subscription to all actorIds (fanout) vs to a single actor ID?

See https://www.youtube.com/watch?v=8Aj1WUzVvGs&t=1738s

Reply:
> This may not be as bad, but we do have generally a large number of pods and so the mapping tables could get large. 

The same way this is handled with proxying existing requests today.

Reply:
Maybe a diagram can help explain this better. Each sidecar instance will read from the message broker, so will they compete for messages and redirect to the right sidecar for each actor id?